platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/CloseableServiceRegistry.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service;
+
+import java.io.Closeable;
+
+public interface CloseableServiceRegistry extends ServiceRegistry, Closeable {
```

Comment:
💭  In many cases, the code that created a service registry didn't care that it was `DefaultServiceRegistry` specifically. However, at the boundaries between the scopes (like Build Session, Build Tree, etc) the state management needed the ability to close the children registries when the parent is being closed.